tests/rados/test_pool_osd_recovery.py

```diff
@@ -0,0 +1,348 @@
+from tests.rados.rados_test_util import get_device_path, wait_for_device
+from tests.rados.stretch_cluster import get_osd_details, wait_for_clean_pg_sets
+from tests.rados.test_data_migration_bw_pools import create_given_pool
+from utility.log import Log
+from utility.utils import method_should_succeed
+
+log = Log(__name__)
+
+
+def run(ceph_cluster, **kw) -> int:
+    """
+    Test to perform tier-4 tests on pools related to OSD daemons
+    Returns:
+        1 -> Fail, 0 -> Pass
+    """
+    log.info(run.__doc__)
+    config = kw["config"]
+    cephadm = CephAdmin(cluster=ceph_cluster, **config)
+    rados_obj = RadosOrchestrator(node=cephadm)
+    pool_obj = PoolFunctions(node=cephadm)
+    pool_configs = config["pool_configs"]
+    pool_configs_path = config["pool_configs_path"]
+    osd_nodes = ceph_cluster.get_nodes(role="osd")
+
+    with open(pool_configs_path, "r") as fd:
+        pool_conf_file = yaml.safe_load(fd)
+
+    pools = []
+    for i in pool_configs:
+        pool = pool_conf_file[i["type"]][i["conf"]]
+        create_given_pool(rados_obj, pool)
+        pools.append(pool["pool_name"])
+
+    log.info(f"Created {len(pools)} pools for testing. pools : {pools}")
+
+    # Checking cluster health before starting the tests
+    method_should_succeed(rados_obj.run_pool_sanity_check)
+    log.info(
+        "Completed health cehck of the cluster after test pool creation. Cluster health good!!"
+    )
+
+    log.info("\n\n---- Starting All workflows ----\n\n")
+    # rebooting one OSD and checking cluster health
+    log.info(
+        "---- Starting workflow ----\n---- 1. Reboot of single osd and health check ----"
+    )
+    for pool in pools:
+        pool_id = pool_obj.get_pool_id(pool_name=pool)
+        pgid = f"{pool_id}.0"
+        pg_set = rados_obj.get_pg_acting_set(pg_num=pgid)
+        target_osd = pg_set[0]
+        log.debug(f"Rebooting OSD : {target_osd} and checking health status")
+        if not rados_obj.change_osd_state(action="restart", target=target_osd):
+            log.error(f"Unable to restart the OSD : {target_osd}")
+            raise Exception("Execution error")
+
+        # Waiting for recovery to post OSD reboot
+        method_should_succeed(wait_for_clean_pg_sets, rados_obj)
+        log.debug(
+            "PG's are active + clean post OSD reboot, proceeding to restart next OSD"
+        )
+
+    log.info("All the planned primary OSD reboots have completed")
+    # Checking cluster health after the tests
+    method_should_succeed(rados_obj.run_pool_sanity_check)
+    log.info("----Completed workflows 1. Reboot of single osd and health check----")
+
+    log.info("---- Starting workflow ----\n---- 2. Rolling Reboot OSD hosts ----")
+    for nodes in osd_nodes:
+        log.info(f"Rebooting node : {nodes.hostname}")
+        nodes.exec_command(sudo=True, cmd="reboot", long_running=True)
+
+        # Sleeping for 10 seconds and starting verification.
+        time.sleep(10)
```

Comment:
IMO below method waits for recovery to complete. can we remove time.sleep above?

Reply:
The above method does not wait for recovery.

Reply:
Once the nodes are rebooted, I have observed that it takes some time to start balancing, identifying that nodes have gone down.

This 10 seconds helps achieve that. So cannot remove the 10 seconds. Can be reduced, but I have seen that 10 secs should cover all.